Implementation of Non-Gaussian transition models

zhenyuen opened the pull request:
@hpritchett-dstl 

Still W.I.P.

Features:

1. Introduce a new family of transition models that generalise transition models driven by white noise to the Levy process.
2. Provides support for normal sigma-mean and normal variance-mean mixture processes.
3. Users can inject the desired driving processes into the selected transition model
4. Introduce marginalised particle filtering for inferencing the newly-introduced class of models.

Things to do:

1. Update tutorial notebook for the new class of models and marginalised particle filter
2. Writing unit tests, still figuring out how to test every possible combination of the driving process and transition model pair without bloat, suggestions are welcome.
3. Implementation of PDF and log PDF functions, certain Levy distributions have non-trivial PDFs/CDFs, currently looking into inverse FFT methods involving characteristic functions.
4. Provide support for more models, e.g. Singer

Additional comments:

I believe the current design is a good compromise between speed and compatibility with the existing stone soup framework to avoid introducing any breaking changes. The newly introduced models decouple the deterministic (transition model) and non-deterministic (noise driver) components. That being said, suggestions are welcome and highly appreciated.

hpritchett-dstl commented on docs/tutorials/11_LevyTransitionModels_&_MarginalisedParticleFilter.ipynb:
Notebook looks great, but it needs to be converted into a python file. See the other tutorials for the formatting you need to follow. I also think this tutorial should be in the tutorials/filters directory.

hpritchett-dstl commented on stonesoup/models/tests/test_driver.py:
```diff
@@ -0,0 +1,194 @@
+        assert((results.pvalue >= threshold) == ks_result) # 99% CI
+ 
+
+# def test_ts_nvm_residuals():
```
remove commented out test.

hpritchett-dstl commented on stonesoup/types/state.py:
```diff
@@ -829,6 +829,53 @@ def __getitem__(self, item):
+        mu_bar = np.sum(self.weight[np.newaxis, :] * mu, axis=1)
+        tmp = mu - mu_bar[:, np.newaxis]
+        weighted_mean = np.sum(self.weight[np.newaxis, np.newaxis, :] * (np.einsum("ik,jk->ijk", tmp, tmp)), axis=len(covariance.shape)-1)
+        # print("In")
+        # print(weighted_mean + weighted_covar )
+        # print("In")
+        # print(covariance)
```
remove commented out code used for testing

hpritchett-dstl commented on stonesoup/types/state.py:
```diff
@@ -829,6 +829,53 @@ def __getitem__(self, item):
+            return self.state_vector
+        return np.average(self.state_vector, axis=1, weights=np.exp(self.log_weight))
+
+    # @clearable_cached_property('state_vector', 'log_weight', 'covariance')
```
remove commented out code

hpritchett-dstl commented on stonesoup/updater/particle.py:
```diff
@@ -651,3 +652,116 @@ def _get_measurement_loglikelihoods(self, prediction, detections):
         return g
+
+
+class MarginalisedParticleUpdater(ParticleUpdater):
```
missing docstring

hpritchett-dstl commented on stonesoup/updater/particle.py:
```diff
@@ -651,3 +652,116 @@ def _get_measurement_loglikelihoods(self, prediction, detections):
+    def _posterior_covariance(self, hypothesis):
+        predicted_covar = hypothesis.prediction.covariance
+        post_cov = np.zeros_like(predicted_covar)
+        # num_samples = post_cov.shape[-1]
+        # for p in range(num_samples):
+        #     mp_covar = hypothesis.measurement_prediction.covariance[..., p] # M x M X N
+        #     mp_cross_covar =  hypothesis.measurement_prediction.cross_covar[..., p]
+            
+        #     kalman_gain = mp_cross_covar @ np.linalg.inv(mp_covar)
+        #     post_cov[..., p] = predicted_covar[..., p] - kalman_gain @ mp_covar @ kalman_gain.T
```
remove commented out code

hpritchett-dstl commented on stonesoup/predictor/particle.py:
```diff
@@ -525,3 +525,77 @@ def predict(self, prior, timestamp=None, **kwargs):
+        process_covar = process_covar.T# (M, M, N)
+
+
+        # Debug for debugging
```
remove commented out code used for debugging

hpritchett-dstl commented on stonesoup/models/transition/tests/test_ca.py:
```diff
@@ -140,3 +152,119 @@ def base(state_vec, noise_diff_coeffs):
+    mat_list = [base_mat for num in range(0, dim)]
+    F = sp.linalg.block_diag(*mat_list)
+
+    # base_covar = np.array([[timediff**5 / 20,
```
remove commented out code in this function

zhenyuen commented:
@hpritchett-dstl 
I believe I’ve addressed all your comments and refactored the code accordingly. However, my tests on CircleCI for Python 3.8 seem to be failing due to the KS test from `SciPy`. I’m not entirely sure what’s causing this, so I’d appreciate it if you could take a look and help troubleshoot.

Additionally, I noticed that some of my formatting changes might have affected other parts of the repository that aren’t mine. I’ll need more time to undo those, but it might be helpful if we had standardized code styles throughout the repository. For example:

1. Using absolute imports instead of relative imports
2. Configuring `flake8` and `black` consistently, ideally with a pre-commit hook

Lastly, it would be really helpful to have a local development setup that mirrors the CircleCI test environment. This way, I could run the tests locally and address any issues before pushing changes. A docker image would be great!

nperree-dstl commented on docs/tutorials/11_LevyTransitionModels_&_MarginalisedParticleFilter.py:
```diff
@@ -0,0 +1,300 @@
+# 
+# 
+# 
+# <!-- and  in the $\alpha$-stable law is a function of the conditional Gaussian mean $\mu_W$
+# 
+# where $\beta=\begin{cases} 1, \quad \mu_W \neq 0 \\ 0, \quad \text{otherwise} \end{cases}$ with $\beta=0$ being the a symmetric stable distribution.
+# 
+# $\sigma=\frac{\mathbb{E}|w|^\alpha \Gamma(2-\alpha) \cos(\pi \alpha / 2))}{1- \alpha}$ represent the scale parameter and $\beta=$ controlling the skewness of the stable distribution. -->
+# 
```
Does this part need to be removed?

zhenyuen replied:
Thanks for pointing that out, will do.

nperree-dstl commented on stonesoup/models/base_driver.py:
```diff
@@ -0,0 +1,426 @@
+from abc import abstractmethod
```
This file (stonesoup.models.base_driver) needs to be added to the docs/source/stonesoup.models.rst file for it to appear in the documentation. It may be worth creating a subfolder for driver models similar to clutter models or transition models.

zhenyuen replied:
As the Gaussian model lives within `models.base`, I created a separate `LevyModel` class as it should be as general as the Gaussian model, with a similar interface. Would it help if I created a separate `stonesoup/models/driver/` folder to store the implementation of the drivers there? But I leave the `base_driver.py` under `stonesoup/models/`

nperree-dstl commented on stonesoup/models/driver.py:
```diff
@@ -0,0 +1,251 @@
+from typing import Callable, Generator, Optional, Union
```
This file (stonesoup.models.driver) needs to be added to the docs/source/stonesoup.models.rst file for it to appear in the documentation.

nperree-dstl reviewed (changes requested):
I've made some comments on the tutorial that could do with being applied throughout to make the documentation render properly. There is some guidance here which might be useful: https://sphinx-gallery.github.io/stable/syntax.html

zhenyuen commented on docs/tutorials/11_LevyTransitionModels_&_MarginalisedParticleFilter.py:
```diff
@@ -0,0 +1,300 @@
+from datetime import datetime, timedelta
+
+
+# The state of the target can be represented as 2D Cartesian coordinates, $\left[x, \dot x, y, \dot y\right]^{\top}$, modelling both its position and velocity. A simple truth path is created with a sampling rate of 1 Hz.
+
+# In[2]:
```
> you can also render the docs locally by running python -m sphinx -W docs/source docs/build.

Thanks for that, will try. That being said, running `-W` fails as there appears as warnings from other sources are being treated as errors. For example,

```
generating gallery...
generating gallery for auto_tutorials... [  9%] 01_KalmanFilterTutorial.py
Warning, treated as error:

../tutorials/01_KalmanFilterTutorial.py unexpectedly failed to execute correctly:

    Traceback (most recent call last):
      File "/Users/chongzhenyuen/Documents/iib-project/stonesoup/docs/tutorials/01_KalmanFilterTutorial.py", line 192, in <module>
        plotter.plot_ground_truths(truth, [0, 2])
      File "/Users/chongzhenyuen/Documents/iib-project/stonesoup/stonesoup/plotter.py", line 2495, in plot_ground_truths
        self.fig.add_trace(go.Scatter(truth_kwargs))
      File "/Users/chongzhenyuen/miniforge3/envs/iib-project/lib/python3.10/site-packages/plotly/graph_objs/_scatter.py", line 3192, in __init__
        self._process_kwargs(**dict(arg, **kwargs))
      File "/Users/chongzhenyuen/miniforge3/envs/iib-project/lib/python3.10/site-packages/plotly/basedatatypes.py", line 4322, in _process_kwargs
        raise err
    ValueError: Invalid property specified for object of type plotly.graph_objs.Scatter: 'legendrank'

    Did you mean "legendgroup"?
```

pumpkinboo commented on docs/tutorials/11_LevyTransitionModels_&_MarginalisedParticleFilter.py:
```diff
@@ -0,0 +1,300 @@
+
+# Consider the scenario where the target evolves according to the Langevin model, driven by a normal sigma-mean mixture with the mixing distribution being the $\alpha$-stable distribution.
+
+# In[1]:
```
Hey @nperree-dstl , sorry for the absence in the past few months. I have just committed the fixes for this. Let me know if anything else is required, thanks!

zhenyuen replied:
Hey @nperree-dstl , sorry for the few months of absence. I have just committed the changes to fix this. Let me know if there is anything else, thanks!
